### PR TITLE
Adds a debug trait that enables and gives 100% success chance on manual bionic installs

### PR DESCRIPTION
**Pull request by ghost**
The debug trait lets you install bionics manually, and with a 100% success rate. Given that's debug, of course, it's not meant for normal play, and is mainly to make testing new bionics painless.

# This PR has been changed! You can see the old description below, to give context on the conversation:

I had this idea some time ago, and it was pretty easy to pull off!

This PR adds the Fluid Pumps CBM. Parts of your inner palm have been replaced with grated piping that lets you suck up liquids from the ground; a handy access hatch where your belly button used to be can be used to catch the liquid as you suck it up. It has a difficulty of 2, costs 2 power to activate, and consumes an additional 1 power every 10 turns while active

Codewise, I modified how the game handles liquids that you can't pick up. If you have Fluid Pumps active, instead of throwing the message `You can't pick up a liquid!`, the game gets that liquid from the tile and prompts the player to handle it like it was in a container such as a toilet cistern.

Fluid Pumps can be found alongside normal CBMs, as well as from butchering technicians, who could get use out of it pre-Cataclysm by carrying a tank or jerrycan with them and siphoning chemical spills or caustic materials from the ground as they go.

### CODE REVIEW

**Comment by MT-Arnoldussen** [comment]
Just FYI, I was just reading the [frequently made suggestions](https://discourse.cataclysmdda.org/t/fms-frequently-made-suggestions/9764), and it specifically mentions 

> Recover liquids from the ground: No interest/wouldn’t work well
This keeps being suggested, but it’s just not reasonable to recover spilled liquids from the ground and then get any kind of use out of them, since in general they’d be so adulterated by whatever would get mixed in with them that they wouldn’t be fit for any purpose.

Of course, this being done with a CBM might make it more realistic.

**Comment by Xhuis** [comment]
Oh, I didn't see that. If Kevin has a problem with it, I can close this PR. I think as a bionic it *does* make feasible sense, however.

**Comment by kevingranade** [comment]
I dont see how making it a CBM improves things.  You just spilled a bunch of gasoline or milk or water in some dirt, how do you get it back out?

**Comment by Xhuis** [comment]
I personally see it as a thin mesh or sieve over the mouth of the pipe. Pressing your hands to the ground will suck up the wet ground, and the sieve will separate the liquid from the solid, allowing you to suck it up from where it was spilled.

There are bionics such as the blood filter that function at a microscopic level to intelligently separate blood from foreign substances, the monomolecular blade which is sharpened to a single-atom point, and a facial distortion unit that can dynamically make very tiny adjustments to your face to make you a better conversationalist. I think it's certainly technologically feasible, if we wanted to go that route.

**Comment by sfsworms** [comment]
What about only part of the liquid being recoverable?

**Comment by kevingranade** [comment]
tl;dr
There are some niche situations where liquid retrieval from the ground would be a reasonable thing to do. Instead of just declaring that there's a magical solution for all the problems inherent in the situation, figure out whether you can identify those niche situations and support just those.  For example, you could mop up water-based liquids from a hard surface and retrieve it as "dirty water", that can be an input for crafting recipes to try and recover some of it.  Potentially the same can be applied to other liquids, though the only one that comes to mind that would be worthwhile is gasoline.

Also see #9461, the tl;dr of which is, "only add this if you're willing to do all the work to make it sensical".

> Pressing your hands to the ground will suck up the wet ground

I assume you mean, "will suck it off of the wet ground" based on the previous description, something like a vacuum tool that handles liquids.  This ignores "wetting" of the dirt particles, simple suction or absorption won't retrieve all of the liquid, because it's bound tightly to individual dirt particles by surface tension. Take a look at a used mop bucket or a wet/dry vac that has been used to pick up water off a floor.  Your water is in there, but it's pretty hopelessly adulterated with gross stuff from the floor, and that's from a relatively clean floor in a modern home, not arbitrarally dirty environments you're going to be in in the game.  Basically, using airflow to pick it up instead of a sponge or cloth to absorb it makes no difference, it's going to be dirty either way.

> the sieve will separate the liquid from the solid

How does the sieve know what's dirt and what's part of the liquid?  There are quite a number of things we treat as liquids that are actually suspensions of various kinds (blood, tea, coffee, milk), a perfect filter (see reverse osmosis) is going to just give you water from most of those, and a non-perfect filter is going to adulterate your liquid with a varying amount of dirt.

Even outside of suspensions, as soon as your liquid hits the ground it's going to start dissolving things it comes into contact with, so it becomes a whole solution and it's now adulterated with new materials, and your filter is going to definitely be unable to separate them.

> I think the main issue is that fluids vary wildly in how realistic it would be to suck them up, even if you have some ideal filter system.

So much this, there are some vaguely reasonable scenarios, but int the vast majority of situations you're going to encounter, the liquid is for all intents and purposes destroyed as soon as it hits the ground.

> Ideally water spilled on dirt should probably not exist after a couple minutes in-game

I agree, adding a permeable tag would probably help settle this recurring issue because it would make it more clear that the liquid is gone.

> What about only part of the liquid being recoverable?

How much?  The problem is it varies wildly depending on the situation, so there's no consistent amount to retrieve (other than "none").

> such as the blood filter that function at a microscopic level to intelligently separate blood from foreign substances

The problem with this as an example is that it has a known quantity to operate on, blood, and generally known contaminants to separate, so it can be designed to do this very precisely.  With the liquid retrieval unit, it's attempting to separate unknown contaminants from unknown liquids, so in addition to the task of actually handling the separation, it also has to magically know what kind of liquid it's dealing with.

Aside from that, the "we have X so we should have Y" argument is, in general, invalid.  Some pre-existing items in the game are legacy items that I would never allow now, but it's not worth it to remove them (though every time someone uses this argument it makes it more worthwhile).

**Comment by Xhuis** [comment]
hmmm... they're reasonable points. While I do still think that "we have X so Y is more reasonable" is a valid point, as it helps gauge the level of technology the world has developed to, in this case it may be more complex.

I'm willing to make this work, however. My initial thoughts would be to give certain terrain - such as tiles, ceramic, maybe roads, etc - a new flag called `HYDROPHOBIC` that would allow liquids spilled on them to be sucked up using this new CBM, easily. Spilling some water on a tiled floor would be easy to clean up in real-life, so I don't see why it couldn't work here. Other terrain, like grass, would instead have a flag called `ABSORBENT` that would cause liquids spilt on them to despawn after some time.

Realistically speaking, the thought of many liquids being easier to suck up than others is also a valid argument. Spilling a bottle of water would be much easier to clean than spilling a bottle of syrup, because the syrup is sticky and once it's spilled it wouldn't like coming back up. For this, I'd think a tag such as `PERMANENT_SPILL` for these liquids would work, which would prevent them from being retrieved after they're spilled.

If we wanted to go more in-depth with it, we could add a new json field `retrievability` or something like that, and only liquid above the amount could be retrieved; ala if clean water has a retrievability of 1, and there's 4 spilt clean water, 3 could be retrieved, but the 1 remaining could not.

I feel like this would help maintain the purpose of the CBM and let it still have a use without compromising the issue into "no liquid retrieval" or "all liquid retrieval." A simple sieve/filter would be able to separate small dirt particles from liquid spilled onto surfaces that are already clean and solid such as tiles, which makes the technology behind it more plausible.

Would that work?

**Comment by DracoGriffin** [comment]
> joke about the five-second rule

Would end up being thirty-second rule as five turns is 6 seconds each, but funny nonetheless. I like the idea of `ABSORBENT` and `HYDROPHOBIC` flags, may be needed to change names but useful start. Potentially either crowdsource some feedback in a separate issue with your ideas on what liquids/tiles should function in this new way, or if you have the time and idea already, perhaps a PR with a decent amount of it already functionally applied as a jumping off point.

Either way, I applaud your efforts for taking on basically a four year old issue with the appropriate nuance and levelheadedness.

**Comment by mlangsdorf** [comment]
Thank you so much for handling the debug install of CBMs.

**Comment by kevingranade** [comment]
What is the pre-cataclysm purpose of the CBM anyway?  Your stated scenario doesn't require any of the actually advanced parts of the proposal, just something that can retrieve liquids from the ground, but it wouldn't matter if those liquids were contaminated, so why does it have the presumably very advanced filtering function built in?  For that matter, why is it a CBM in the first place, since it's a glorified wet-vac?

If you want potable liquids, there are a million times easier ways to get them (even post cataclysm) than a magic filtering appendage, so why would anyone ever build the magical filtering appendage?  Yes we have silly things like a corpse dehydrator, but that again is something that was added for cool points with no critical thinking applied, and is another example of something I would turn down immediately if suggested today.

How about this, drop the CBM, adjust the mop code to retrieve water-based liquids from hard surfaces (needs that HYDROPHOBIC flag, or how about NONPOROUS instead) and deposit them in a bucket or other container as "dirty water", and add recipes that let you filter "dirty water" to get "water", which can then be purified or whatever as usual.

Optionally: Make liquids on non-NONPOROUS terrain dissapear more quickly.
Other liquids?  I do not see an approach that would work for gasoline or other chemicals that's worth pursuing, but I might be missing something.

**Comment by Xhuis** [comment]
A large shift in PR scale is something I'm not up to doing tonight due to a frustrating day at work and little time for the night, so I suppose I'll scrap the CBM and leave in the debug trait for a standalone merge.

**Comment by Xhuis** [comment]
Should be good to go now.
